File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/appleskin/AppleSkinSyncTask.java
package crabcraft.net.crabUtilities.appleskin;

import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.nio.ByteBuffer;

final class AppleSkinSyncTask extends BukkitRunnable {

    private static final float MINIMUM_EXHAUSTION_CHANGE = 0.01F;

    private final Player player;
    private final long generation;
    private float previousSaturation = -1;
    private float previousExhaustion = -1;

    AppleSkinSyncTask(Player player) {
        this.player = player;
        this.generation = AppleSkinIntegration.currentGeneration();
        runTaskTimer(AppleSkinIntegration.plugin(), 1, 1);
    }

    @Override
    public void run() {
        if (!AppleSkinIntegration.isCurrentGeneration(generation) || !player.isOnline()) {
            cancel();
            return;
        }

        final float saturation = player.getSaturation();
        if (saturation != previousSaturation) {
            send(AppleSkinIntegration.SATURATION_CHANNEL, saturation);
            previousSaturation = saturation;
        }

        final float exhaustion = player.getExhaustion();
        if (Math.abs(exhaustion - previousExhaustion) >= MINIMUM_EXHAUSTION_CHANGE) {
            send(AppleSkinIntegration.EXHAUSTION_CHANNEL, exhaustion);
            previousExhaustion = exhaustion;
        }
    }

    private void send(String channel, float value) {
        player.sendPluginMessage(
                AppleSkinIntegration.plugin(),
                channel,
                ByteBuffer.allocate(Float.BYTES).putFloat(value).array());
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/appleskin/AppleSkinSyncTask.kt
package crabcraft.net.crabUtilities.appleskin

import java.nio.ByteBuffer
import kotlin.math.abs
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable

class AppleSkinSyncTask(private val player: Player) : BukkitRunnable() {
    private val generation = AppleSkinIntegration.currentGeneration()
    private var previousSaturation = -1f
    private var previousExhaustion = -1f

    init {
        runTaskTimer(AppleSkinIntegration.plugin(), 1, 1)
    }

    override fun run() {
        if (!AppleSkinIntegration.isCurrentGeneration(generation) || !player.isOnline) {
            cancel()
            return
        }
        val saturation = player.saturation
        if (saturation != previousSaturation) {
            send(AppleSkinIntegration.SATURATION_CHANNEL, saturation)
            previousSaturation = saturation
        }
        val exhaustion = player.exhaustion
        if (abs(exhaustion - previousExhaustion) >= MINIMUM_EXHAUSTION_CHANGE) {
            send(AppleSkinIntegration.EXHAUSTION_CHANNEL, exhaustion)
            previousExhaustion = exhaustion
        }
    }

    private fun send(channel: String, value: Float) {
        player.sendPluginMessage(
            AppleSkinIntegration.plugin(),
            channel,
            ByteBuffer.allocate(Float.SIZE_BYTES).putFloat(value).array(),
        )
    }

    companion object {
        private const val MINIMUM_EXHAUSTION_CHANGE = 0.01f
    }
}
